theseus/labs/lie/lie_tensor.py

```diff
@@ -0,0 +1,146 @@
+        return LieTensor(t, ltype=self.ltype)
+
+    # Operators
+    def exp(self, tangent_vector: torch.Tensor) -> "LieTensor":
+        return self.new(self._fn_lib.exp(tangent_vector))
+
+    def log(self, group: "LieTensor") -> torch.Tensor:
+        return self._fn_lib.log(group._t)
+
+    def adj(self, group: "LieTensor") -> "LieTensor":
+        return self.new(self._fn_lib.adj(group._t))
+
+    def inv(self, group: "LieTensor") -> "LieTensor":
+        return self.new(self._fn_lib.inv(group._t))
+
+    def hat(self, tangent_vector: torch.Tensor) -> torch.Tensor:
+        return self._fn_lib.hat(tangent_vector)
+
+    def vee(self, matrix: torch.Tensor) -> torch.Tensor:
+        return self._fn_lib.vee(matrix)
+
+    def lift(self, matrix: torch.Tensor) -> torch.Tensor:
+        return self._fn_lib.lift(matrix)
+
+    def project(self, matrix: torch.Tensor) -> torch.Tensor:
+        return self._fn_lib.project(matrix)
+
+    def compose(self, other: "LieTensor") -> "LieTensor":
+        self._check_ltype(other, "compose")
+        return self.new(self._fn_lib.compose(self._t, other._t))
+
+    def left_act(self, matrix: torch.Tensor) -> torch.Tensor:
+        return self._fn_lib.left_act(self._t, matrix)
+
+    def left_project(self, matrix: torch.Tensor) -> torch.Tensor:
+        return self._fn_lib.left_project(self._t, matrix)
+
+    # Operator Jacobians
+    def _unary_jop_base(
+        self,
+        input0: torch.Tensor,
+        fn: UnaryOperatorOpFnType,
+        out_is_group: bool = True,
+    ) -> _JFnReturnType:
+        jacs: List[torch.Tensor] = []
+        op_res: TensorType = fn(input0, jacobians=jacs)
+        if out_is_group:
+            op_res = self.new(op_res)
+        return jacs, op_res
+
+    def jexp(self, tangent_vector: torch.Tensor) -> _JFnReturnType:
+        return self._unary_jop_base(tangent_vector, self._fn_lib.exp)
+
+    def jlog(self, group: "LieTensor") -> _JFnReturnType:
+        return self._unary_jop_base(group._t, self._fn_lib.exp, out_is_group=False)
+
+    def jinv(self, group: "LieTensor") -> _JFnReturnType:
+        return self._unary_jop_base(group._t, self._fn_lib.exp)
+
+    def jcompose(self, other: "LieTensor") -> _JFnReturnType:
+        self._check_ltype(other, "jcompose")
+        jacs: List[torch.Tensor] = []
+        op_res = self.new(self._fn_lib.compose(self._t, other._t, jacobians=jacs))
+        return jacs, op_res
+
+    def _no_unary_op(self, input0: TensorType) -> _JFnReturnType:
+        raise NotImplementedError
+
+    def _no_binary_op(self, input0: TensorType, input1: TensorType) -> _JFnReturnType:
+        raise NotImplementedError
+
+    jadjoint = _no_unary_op
+    jhat = _no_unary_op
+    jvee = _no_unary_op
+    jlift = _no_unary_op
+    jproject = _no_unary_op
+    jleft_act = _no_binary_op
+    jleft_project = _no_binary_op
+
+
+# ----------------------------
+# Tensor creation functions
+# ----------------------------
+def new(
+    data: Any,
```

Comment:
Any -> TensorType?

Reply:
This is what torch uses because `torch.as_tensor` can receive float, numpy array, sequences of many different types, torch tensors, and who knows what else. Handling types for this is probably a big mess.

Reply:
Ah okay, makes sense.